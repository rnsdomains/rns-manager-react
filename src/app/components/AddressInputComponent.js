import React, { useState } from 'react';
import propTypes from 'prop-types';
import { Button } from 'react-bootstrap';
import { toChecksumAddress } from 'rskjs-util';

import { validateAddress } from '../validations';
import { ChecksumErrorContainer } from '../containers';
import UserErrorComponent from './UserErrorComponent';
import UserSuccessComponent from './UserSuccessComponent';
import UserWaitingComponent from './UserWaitingComponent';

import edit from '../../assets/img/edit.svg';
import editActive from '../../assets/img/edit-active.svg';
import closeBlue from '../../assets/img/close-blue.svg';

const AddressInputComponent = ({
  allowDelete,
  label,
  labelDisplay,
  labelIcon,
  value,
  isWaiting,
  isError,
  handleErrorClose,
  handleSuccessClose,
  handleSubmit,
  handleDelete,
  isSuccess,
  strings,
  successTx,
  validationChainId,
  validation,
  suggestions,
}) => {
  const [isEditing, setIsEditing] = useState(false);
  const [isDeleting, setIsDeleting] = useState(false);
  const [isChecksumError, setIsChecksumError] = useState(false);
  const [isLocalError, setIsLocalError] = useState(false);
  const [editText, setEditText] = useState('');

  const handleEditClick = () => {
    setIsEditing(!isEditing);
    setIsDeleting(false);
  };

  const handleDeleteClick = () => {
    setIsDeleting(!isDeleting);
    setIsEditing(false);
  };

  const confirmDelete = () => {
    handleDelete();
  };

  const handleErrorClick = () => {
    setIsLocalError('');
    handleErrorClose();
  };

  const handleSubmitClick = () => {
    if (editText.toLowerCase() === value.toLowerCase()) {
      return setIsLocalError('Value is the same.');
    }

    setIsLocalError(false);
    setIsChecksumError(false);

    if (!validation) {
      return handleSubmit(editText);
    }

    switch (validateAddress(editText, validationChainId)) {
      case 'Invalid address':
        return setIsLocalError('Invalid address');
      case 'Invalid checksum':
        return setIsChecksumError(true);
      default:
    }
    return handleSubmit(editText);
  };

  const handleTextChange = (evt) => {
    setEditText(evt.target.value);
  };

  const handleChecksumClick = () => {
    setEditText(editText.toLowerCase());
    setIsChecksumError(false);
    return handleSubmit(editText);
  };

  const handleCancelClick = () => {
    setIsEditing(false);
    setIsChecksumError(false);
  };

  const formatValue = () => {
    if (value === '') {
      return value;
    }

    return validation ? toChecksumAddress(value, validationChainId) : value;
  };

  if (isSuccess && (isEditing || isDeleting)) {
    setIsEditing(false);
    setIsDeleting(false);
    setEditText('');
  }

  return (
    <div className="row addressInput">
      <div className="row view">
        <div className="col-md-3 label">
          {labelIcon && <img src={labelIcon} alt={labelDisplay || label} />}
          {labelDisplay || label}
        </div>
        <div className={`${allowDelete ? 'col-md-7' : 'col-md-8'} value`}>
          {strings.value_prefix
            && <span className="value-prefix">{`${strings.value_prefix}: `}</span>
          }
          {formatValue()}
        </div>
        <div className={`${allowDelete ? 'col-md-2' : 'col-md-1'} options`}>
          <button
            type="button"
            onClick={handleEditClick}
            className="edit"
            disabled={isWaiting}
          >
            <img src={(!isEditing ? edit : editActive)} alt={strings.edit} />
          </button>
          {allowDelete
            && (
            <button
              type="button"
              onClick={handleDeleteClick}
              className="delete"
              disabled={isWaiting}
            >
              <img src={closeBlue} alt={strings.delete} />
            </button>
            )
          }
        </div>
      </div>
      {isEditing
        && (
        <div className="row edit">
          <div className="col-md-2 editLabel">
            {strings.edit_propmt}
          </div>
          <div className="col-md-6">
            <input
              type="text"
              placeholder={strings.edit_placeholder}
              value={editText}
              onChange={handleTextChange}
              disabled={isWaiting}
            />
          </div>
          <div className="col-md-4 buttons">
            <Button
              className="cancel"
              onClick={() => handleCancelClick()}
              disabled={isWaiting}
            >
              {strings.cancel}
            </Button>
            <Button
              className="submit"
              disabled={isWaiting}
              onClick={handleSubmitClick}
            >
              {strings.submit}
            </Button>
          </div>
          <ul className="suggestions">
            {suggestions.map(item => (
              <li key={item.value}>
                <button
                  type="button"
                  onClick={() => setEditText(item.value)}
                >
                  {item.name}
                </button>
              </li>
            ))}
          </ul>
        </div>
        )
      }

      <ChecksumErrorContainer
        show={isChecksumError}
        inputValue={editText}
        handleClick={() => handleChecksumClick()}
      />

      {isDeleting
        && (
          <div className="delete">
            <p>{strings.delete_confirm_text}</p>
            <p>
              <Button
                className="cancel"
                onClick={() => { setIsDeleting(false); setIsLocalError(''); }}
                disabled={isWaiting}
              >
                {strings.cancel}
              </Button>
              <Button
                className="submit"
                disabled={isWaiting}
                onClick={confirmDelete}
              >
                {strings.delete}
              </Button>
            </p>
          </div>
        )
      }

      <UserWaitingComponent
        message={strings.waiting}
        visible={isWaiting}
      />

      <UserErrorComponent
        title={strings.error_title}
        message={isLocalError || strings.error_message}
        handleCloseClick={() => handleErrorClick()}
        visible={isError || isLocalError}
      />

      <UserSuccessComponent
        title={strings.success_title}
        message={strings.success_message}
        handleCloseClick={handleSuccessClose}
        address={successTx}
        visible={isSuccess}
      />
    </div>
  );
};

AddressInputComponent.defaultProps = {
  allowDelete: true,
  isError: false,
  isWaiting: false,
  isSuccess: false,
  successTx: '',
  validation: true,
  validationChainId: null,
  strings: {
    cancel: 'Cancel',
    delete: 'Delete',
    delete_confirm_text: 'Are you sure you want to delete?',
    edit: 'Edit',
    edit_placeholder: 'enter address',
    edit_propmt: 'Change ownership',
    error_title: 'Error Title',
    error_message: 'Error Message',
    submit: 'Submit',
    success_title: 'Success Title',
    success_message: 'Success Message',
    value_prefix: 'Owner',
    waiting: 'Waiting text',
  },
  handleDelete: () => {},
  handleErrorClose: () => {},
  handleSuccessClose: () => {},
  labelDisplay: null,
  labelIcon: null,
  suggestions: [],
};

AddressInputComponent.propTypes = {
  allowDelete: propTypes.bool,
  label: propTypes.string.isRequired,
  labelDisplay: propTypes.string,
  labelIcon: propTypes.string,
  isError: propTypes.bool,
  isWaiting: propTypes.bool,
  isSuccess: propTypes.bool,
  validation: propTypes.bool,
  successTx: propTypes.string,
  handleErrorClose: propTypes.func,
  handleSuccessClose: propTypes.func,
  handleSubmit: propTypes.func.isRequired,
  handleDelete: propTypes.func,
  value: propTypes.string.isRequired,
  validationChainId: propTypes.string,
  strings: propTypes.shape(),
  suggestions: propTypes.arrayOf(propTypes.shape({
    name: propTypes.string,
    value: propTypes.value,
  })),
};

export default AddressInputComponent;
