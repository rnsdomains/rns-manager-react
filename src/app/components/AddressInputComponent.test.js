import React from 'react';
import { mount } from 'enzyme';

import AddressInputComponent from './AddressInputComponent';

describe('AddressInputComponent', () => {
  const initProps = {
    allowDelete: true,
    label: 'Label',
    value: 'Value',
    handleErrorClose: jest.fn(),
    handleSuccessClose: jest.fn(),
    handleSubmit: jest.fn(),
    handleDelete: jest.fn(),
    strings: {
      cancel: 'cancel string',
      delete: 'delete string',
      delete_confirm_text: 'delete confirm text',
      edit: 'edit text',
      edit_placeholder: 'placeholder text',
      edit_propmt: 'edit prompt text',
      error_title: 'error title text',
      error_message: 'error message text',
      submit: 'submit text',
      success_title: 'success title text',
      success_message: 'success message text',
      value_prefix: 'value prefix',
      waiting: 'waiting text string',
    },
  };

  it('renders without crashing', () => {
    const component = mount(<AddressInputComponent {...initProps} />);
    expect(component).toMatchSnapshot();
  });

  it('opens edit area when edit button is clicked and matches snapshot', () => {
    const component = mount(<AddressInputComponent {...initProps} />);
    component.find('button.edit').simulate('click');
    expect(component.find('div.edit').length).toBe(1);
    expect(component).toMatchSnapshot();
  });

  it('closes the edit area when button or cancel is clicked', () => {
    const component = mount(<AddressInputComponent {...initProps} />);
    component.find('button.edit').simulate('click');
    expect(component.find('div.edit').length).toBe(1);

    // click cancel button:
    component.find('button.cancel').simulate('click');
    expect(component.find('div.edit').length).toBe(0);

    // double click edit button:
    component.find('button.edit').simulate('click').simulate('click');
    expect(component.find('div.edit').length).toBe(0);

    // click the delete button:
    component.find('button.delete').simulate('click');
    expect(component.find('div.edit').length).toBe(0);
  });

  it('shows confirmation window when delete is clicked', () => {
    const component = mount(<AddressInputComponent {...initProps} />);
    component.find('button.delete').simulate('click');
    expect(component.find('div.delete').length).toBe(1);
    expect(component).toMatchSnapshot();
  });

  it('closes delete window when cancel or button is clicked', () => {
    const component = mount(<AddressInputComponent {...initProps} />);
    component.find('button.delete').simulate('click');
    expect(component.find('div.delete').length).toBe(1);

    // click cancel button:
    component.find('button.cancel').simulate('click');
    expect(component.find('div.delete').length).toBe(0);

    // double click cancel button:
    component.find('button.delete').simulate('click').simulate('click');
    expect(component.find('div.delete').length).toBe(0);

    // click edit button:
    component.find('button.edit').simulate('click');
    expect(component.find('div.delete').length).toBe(0);
  });

  it('handles delete function when clicked', () => {
    const component = mount(<AddressInputComponent {...initProps} />);
    component.find('button.delete').simulate('click');
    component.find('button.submit').simulate('click');
    expect(initProps.handleDelete).toBeCalledTimes(1);
  });

  it('delete is not shown when allowDelete is false', () => {
    const localProps = {
      ...initProps,
      allowDelete: false,
    };
    const component = mount(<AddressInputComponent {...localProps} />);
    expect(component.find('button.delete').length).toBe(0);
  });

  it('handles submit function when clicked', () => {
    const localProps = {
      ...initProps,
      value: 'jesse.rsk',
    };

    const component = mount(<AddressInputComponent {...localProps} />);
    component.find('button.edit').simulate('click');
    component.find('button.submit').simulate('click');
    expect(initProps.handleSubmit).toBeCalledTimes(1);
    expect(initProps.handleSubmit).toBeCalledWith('jesse.rsk');
  });

  it('shows waiting div when waiting', () => {
    const localProps = {
      ...initProps,
      isWaiting: true,
    };

    const component = mount(<AddressInputComponent {...localProps} />);
    expect(component.find('div.waiting').length).toBe(1);

    expect(component).toMatchSnapshot();
  });

  it('shows success and handleSuccessClose is called when clicked', () => {
    const localProps = {
      ...initProps,
      onSuccess: true,
    };

    const component = mount(<AddressInputComponent {...localProps} />);
    expect(component.find('div.success').length).toBe(1);

    component.find('div.success').find('button.close').simulate('click');
    expect(initProps.handleSuccessClose).toBeCalledTimes(1);
  });

  it('error is showsn when isError is true', () => {
    const localProps = {
      ...initProps,
      isError: true,
    };

    const component = mount(<AddressInputComponent {...localProps} />);
    expect(component.find('div.error').length).toBe(1);

    component.find('div.error').find('button.close').simulate('click');
    expect(initProps.handleErrorClose).toBeCalledTimes(1);
  });

  it('start, edit, and delete allow custom text ', () => {
    const localProps = {
      ...initProps,
      strings: {
        cancel: 'custom cancel string',
        delete: 'custom delete string',
        delete_confirm_text: 'custom delete confirm text',
        edit: 'custom edit text',
        edit_placeholder: 'custom placeholder text',
        edit_propmt: 'custom edit prompt text',
        submit: 'custom submit text',
        value_prefix: 'custom value prefix',
      },
    };
    const component = mount(<AddressInputComponent {...localProps} />);
    expect(component.find('div.value').text()).toEqual('custom value prefix: Value');

    // edit screen
    component.find('button.edit').simulate('click');
    expect(component.find('div.editLabel').text()).toEqual('custom edit prompt text');
    expect(component.find('button.cancel').text()).toEqual('custom cancel string');
    expect(component.find('button.submit').text()).toEqual('custom submit text');

    // delete screen
    component.find('button.delete').simulate('click');
    expect(component.find('div.delete').find('p').first().text())
      .toEqual('custom delete confirm text');
    expect(component.find('button.cancel').text()).toBe('custom cancel string');
    expect(component.find('button.submit').text()).toBe('custom delete string');
  });

  it('waiting allows cutom text', () => {
    const localProps = {
      ...initProps,
      isWaiting: true,
      strings: {
        waiting: 'custom waiting text string',
      },
    };
    const component = mount(<AddressInputComponent {...localProps} />);
    expect(component.find('div.waiting').text()).toBe('custom waiting text string');
  });

  it('error allows cutom text', () => {
    const localProps = {
      ...initProps,
      isError: true,
      strings: {
        error_title: 'custom error title text',
        error_message: 'custom error message text',
      },
    };
    const component = mount(<AddressInputComponent {...localProps} />);
    expect(component.find('p').at(0).text()).toBe('custom error title text');
    expect(component.find('p').at(1).text()).toBe('custom error message text');
  });

  it('success allows cutom text', () => {
    const localProps = {
      ...initProps,
      onSuccess: true,
      strings: {
        success_title: 'custom success title text',
        success_message: 'custom success message text',
      },
    };
    const component = mount(<AddressInputComponent {...localProps} />);
    expect(component.find('p').at(0).text()).toBe('custom success title text');
    expect(component.find('p').at(1).text()).toBe('custom success message text');
  });
});
