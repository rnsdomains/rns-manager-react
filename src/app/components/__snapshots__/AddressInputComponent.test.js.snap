// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddressInputComponent opens edit area when edit button is clicked and matches snapshot 1`] = `
<AddressInputComponent
  allowDelete={true}
  handleDelete={[MockFunction]}
  handleErrorClose={[MockFunction]}
  handleSubmit={[MockFunction]}
  handleSuccessClose={[MockFunction]}
  isError={false}
  isWaiting={false}
  label="Label"
  onSuccess={false}
  strings={
    Object {
      "cancel": "cancel string",
      "delete": "delete string",
      "delete_confirm_text": "delete confirm text",
      "edit": "edit text",
      "edit_placeholder": "placeholder text",
      "edit_propmt": "edit prompt text",
      "error_message": "error message text",
      "error_title": "error title text",
      "submit": "submit text",
      "success_message": "success message text",
      "success_title": "success title text",
      "value_prefix": "value prefix",
      "waiting": "waiting text string",
    }
  }
  value="Value"
>
  <div
    className="row addressInput"
  >
    <div
      className="row view"
    >
      <div
        className="col-md-3 label"
      >
        Label
      </div>
      <div
        className="col-md-7 value"
      >
        value prefix: 
        Value
      </div>
      <div
        className="col-md-2 options"
      >
        <button
          className="edit"
          onClick={[Function]}
          type="button"
        >
          <img
            alt="edit text"
            src="mock-image"
          />
        </button>
        <button
          className="delete"
          onClick={[Function]}
          type="button"
        >
          <img
            alt="delete string"
            src="mock-image"
          />
        </button>
      </div>
    </div>
    <div
      className="row edit"
    >
      <div
        className="col-md-2 editLabel"
      >
        edit prompt text
      </div>
      <div
        className="col-md-6"
      >
        <input
          disabled={false}
          onChange={[Function]}
          placeholder="placeholder text"
          type="text"
          value="Value"
        />
      </div>
      <div
        className="col-md-4 buttons"
      >
        <Button
          active={false}
          className="cancel"
          disabled={false}
          onClick={[Function]}
          type="button"
          variant="primary"
        >
          <button
            className="cancel btn btn-primary"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            cancel string
          </button>
        </Button>
        <Button
          active={false}
          className="submit"
          disabled={false}
          onClick={[Function]}
          type="button"
          variant="primary"
        >
          <button
            className="submit btn btn-primary"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            submit text
          </button>
        </Button>
      </div>
    </div>
  </div>
</AddressInputComponent>
`;

exports[`AddressInputComponent renders without crashing 1`] = `
<AddressInputComponent
  allowDelete={true}
  handleDelete={[MockFunction]}
  handleErrorClose={[MockFunction]}
  handleSubmit={[MockFunction]}
  handleSuccessClose={[MockFunction]}
  isError={false}
  isWaiting={false}
  label="Label"
  onSuccess={false}
  strings={
    Object {
      "cancel": "cancel string",
      "delete": "delete string",
      "delete_confirm_text": "delete confirm text",
      "edit": "edit text",
      "edit_placeholder": "placeholder text",
      "edit_propmt": "edit prompt text",
      "error_message": "error message text",
      "error_title": "error title text",
      "submit": "submit text",
      "success_message": "success message text",
      "success_title": "success title text",
      "value_prefix": "value prefix",
      "waiting": "waiting text string",
    }
  }
  value="Value"
>
  <div
    className="row addressInput"
  >
    <div
      className="row view"
    >
      <div
        className="col-md-3 label"
      >
        Label
      </div>
      <div
        className="col-md-7 value"
      >
        value prefix: 
        Value
      </div>
      <div
        className="col-md-2 options"
      >
        <button
          className="edit"
          onClick={[Function]}
          type="button"
        >
          <img
            alt="edit text"
            src="mock-image"
          />
        </button>
        <button
          className="delete"
          onClick={[Function]}
          type="button"
        >
          <img
            alt="delete string"
            src="mock-image"
          />
        </button>
      </div>
    </div>
  </div>
</AddressInputComponent>
`;

exports[`AddressInputComponent shows confirmation window when delete is clicked 1`] = `
<AddressInputComponent
  allowDelete={true}
  handleDelete={[MockFunction]}
  handleErrorClose={[MockFunction]}
  handleSubmit={[MockFunction]}
  handleSuccessClose={[MockFunction]}
  isError={false}
  isWaiting={false}
  label="Label"
  onSuccess={false}
  strings={
    Object {
      "cancel": "cancel string",
      "delete": "delete string",
      "delete_confirm_text": "delete confirm text",
      "edit": "edit text",
      "edit_placeholder": "placeholder text",
      "edit_propmt": "edit prompt text",
      "error_message": "error message text",
      "error_title": "error title text",
      "submit": "submit text",
      "success_message": "success message text",
      "success_title": "success title text",
      "value_prefix": "value prefix",
      "waiting": "waiting text string",
    }
  }
  value="Value"
>
  <div
    className="row addressInput"
  >
    <div
      className="row view"
    >
      <div
        className="col-md-3 label"
      >
        Label
      </div>
      <div
        className="col-md-7 value"
      >
        value prefix: 
        Value
      </div>
      <div
        className="col-md-2 options"
      >
        <button
          className="edit"
          onClick={[Function]}
          type="button"
        >
          <img
            alt="edit text"
            src="mock-image"
          />
        </button>
        <button
          className="delete"
          onClick={[Function]}
          type="button"
        >
          <img
            alt="delete string"
            src="mock-image"
          />
        </button>
      </div>
    </div>
    <div
      className="delete"
    >
      <p>
        delete confirm text
      </p>
      <p>
        <Button
          active={false}
          className="cancel"
          disabled={false}
          onClick={[Function]}
          type="button"
          variant="primary"
        >
          <button
            className="cancel btn btn-primary"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            cancel string
          </button>
        </Button>
        <Button
          active={false}
          className="submit"
          disabled={false}
          onClick={[Function]}
          type="button"
          variant="primary"
        >
          <button
            className="submit btn btn-primary"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            delete string
          </button>
        </Button>
      </p>
    </div>
  </div>
</AddressInputComponent>
`;

exports[`AddressInputComponent shows waiting div when waiting 1`] = `
<AddressInputComponent
  allowDelete={true}
  handleDelete={
    [MockFunction] {
      "calls": Array [
        Array [],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  handleErrorClose={[MockFunction]}
  handleSubmit={
    [MockFunction] {
      "calls": Array [
        Array [
          "jesse.rsk",
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  handleSuccessClose={[MockFunction]}
  isError={false}
  isWaiting={true}
  label="Label"
  onSuccess={false}
  strings={
    Object {
      "cancel": "cancel string",
      "delete": "delete string",
      "delete_confirm_text": "delete confirm text",
      "edit": "edit text",
      "edit_placeholder": "placeholder text",
      "edit_propmt": "edit prompt text",
      "error_message": "error message text",
      "error_title": "error title text",
      "submit": "submit text",
      "success_message": "success message text",
      "success_title": "success title text",
      "value_prefix": "value prefix",
      "waiting": "waiting text string",
    }
  }
  value="Value"
>
  <div
    className="row addressInput"
  >
    <div
      className="row view"
    >
      <div
        className="col-md-3 label"
      >
        Label
      </div>
      <div
        className="col-md-7 value"
      >
        value prefix: 
        Value
      </div>
      <div
        className="col-md-2 options"
      >
        <button
          className="edit"
          onClick={[Function]}
          type="button"
        >
          <img
            alt="edit text"
            src="mock-image"
          />
        </button>
        <button
          className="delete"
          onClick={[Function]}
          type="button"
        >
          <img
            alt="delete string"
            src="mock-image"
          />
        </button>
      </div>
    </div>
    <div
      className="row waiting"
    >
      <div
        className="col-md-6 offset-md-2"
      >
        <p>
          waiting text string
        </p>
      </div>
    </div>
  </div>
</AddressInputComponent>
`;
